test-deployment/pom.xml

```diff
@@ -0,0 +1,150 @@
+            <artifactId>weld-se-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml</groupId>
+            <artifactId>aalto-xml</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.codehaus.woodstox</groupId>
+            <artifactId>stax2-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.openjdk.nashorn</groupId>
+            <artifactId>nashorn-core</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <argLine>${surefire.system.args}</argLine>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.wildfly.plugins</groupId>
+                <artifactId>wildfly-maven-plugin</artifactId>
+                <version>${version.org.wildfly.wildfly-maven-plugin}</version>
+                <configuration>
+                    <provisioning-dir>${jboss.home}</provisioning-dir>
+                    <offline>${galleon.offline}</offline>
+                    <galleon-options>
+                        <jboss-fork-embedded>${galleon.fork.embedded}</jboss-fork-embedded>
+                    </galleon-options>
+                    <feature-packs>
+                        <feature-pack>
+                            <groupId>org.wildfly</groupId>
+                            <artifactId>wildfly-ee-galleon-pack</artifactId>
+                            <version>${version.org.wildfly}</version>
+                            <inherit-configs>false</inherit-configs>
+                            <included-configs>
+                                <config>
+                                    <model>standalone</model>
+                                    <name>standalone-full.xml</name>
+                                </config>
+                                <config>
+                                    <model>standalone</model>
+                                    <name>standalone.xml</name>
+                                </config>
+                            </included-configs>
+                            <excluded-packages>
+                                <name>docs.schema</name>
+                                <name>appclient</name>
+                                <name>domain</name>
+                            </excluded-packages>
+                        </feature-pack>
+                        <feature-pack>
```

Comment:
@jamezp I'm trying to test with the generated `feature-pack` and it failed with following message:

```bash
➤ mvn wildfly:package                                                                                                                                                                                                                                                                       00:13:40
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.jberet.test:test-deployment >-------------------
[INFO] Building test-deployment 2.1.4-SNAPSHOT
[INFO] --------------------------------[ war ]---------------------------------
[INFO] 
[INFO] --- wildfly-maven-plugin:4.2.1.Final:package (default-cli) @ test-deployment ---
[INFO] Provisioning server in /Users/weli/works/jsr352/test-deployment/target/test-server/wildfly-30.0.0.Final
[INFO] Resolving feature-packs
[INFO] Installing packages
[INFO] Resolving artifacts
[INFO] Generating configurations
[INFO] Delayed generation, waiting...
Forked embedded process has failed with the following error:
org.wildfly.galleon.plugin.server.ConfigGeneratorException: Failed to generate standalone.xml on {
    "operation" => "composite",
    "address" => [],
    "rollback-on-runtime-failure" => true,
    "steps" => [
        {
            "operation" => "add",
            "address" => [{"interface" => "public"}],
            "inet-address" => "${jboss.bind.address:127.0.0.1}"
        },
        {
            "operation" => "add",
            "address" => [{"interface" => "management"}],
            "inet-address" => "${jboss.bind.address.management:127.0.0.1}"
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.naming"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.ee"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.deployment-scanner"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.io"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.undertow"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.elytron"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.jmx"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.remoting"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.logging"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.core-management"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.request-controller"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.security.manager"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.transactions"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.connector"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.jaxrs"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.weld"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.bean-validation"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.jpa"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.clustering.infinispan"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.ee-security"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.metrics"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.health"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.elytron-oidc-client"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.ejb3"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.clustering.ejb"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.discovery"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.clustering.web"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.jdr"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.jsf"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.webservices"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.pojo"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.mail"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.jboss.as.sar"}]
        },
        {
            "operation" => "add",
            "address" => [{"extension" => "org.wildfly.extension.batch.jberet"}]
        }
    ]
}: {"WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:" => {"Operation step-36" => "WFLYCTL0158: Operation handler failed: java.lang.RuntimeException: WFLYCTL0311: Failed to load Extension module org.wildfly.extension.batch.jberet"}}
        at org.wildfly.galleon.plugin.config.generator.BaseConfigGenerator.doHandle(BaseConfigGenerator.java:144)
        at org.wildfly.galleon.plugin.config.generator.BaseConfigGenerator.endBatch(BaseConfigGenerator.java:162)
        at org.wildfly.galleon.plugin.config.generator.ForkedConfigGenerator.executeScript(ForkedConfigGenerator.java:63)
        at org.wildfly.galleon.plugin.config.generator.ForkedConfigGenerator.forkedForEmbedded(ForkedConfigGenerator.java:45)
        at org.wildfly.galleon.plugin.server.ForkedProcessRunner.main(ForkedProcessRunner.java:56)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.066 s
[INFO] Finished at: 2023-12-02T00:21:54+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:4.2.1.Final:package (default-cli) on project test-deployment: Provisioning failed: Forked embedded process has failed -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

It seems the `org.wildfly.extension.batch.jberet` inside the provisioned WildFly module has conflict with the produced feature pack.

I have checked the `module.xml` of `org.wildfly.extension.batch.jberet` and it's like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!--
  ~ Copyright The WildFly Authors
  ~ SPDX-License-Identifier: Apache-2.0
  -->
<module name="org.wildfly.extension.batch.jberet" xmlns="urn:jboss:module:1.9">

    <properties>
        <property name="jboss.api" value="private"/>
    </properties>

    <exports>
        <exclude path="org/wildfly/extension/batch/jberet/_private"/>
    </exports>

    <resources>
        <resource-root path="wildfly-batch-jberet-30.0.0.Final.jar"/>
    </resources>

    <dependencies>
        <module name="java.sql"/>
        <module name="jakarta.batch.api"/>
        <module name="jakarta.enterprise.api"/>
        <module name="jakarta.transaction.api"/>
        <module name="org.jberet.jberet-core" services="import"/>
        <module name="org.jboss.as.controller"/>
        <module name="org.jboss.as.ee"/>
        <module name="org.jboss.as.naming"/>
        <module name="org.jboss.as.threads"/>
        <module name="org.jboss.as.server"/>
        <!-- Only used if capability org.wildfly.weld is available -->
        <module name="org.jboss.as.weld.common" optional="true"/>
        <module name="org.jboss.logging"/>
        <module name="org.jboss.modules"/>
        <module name="org.jboss.msc"/>
        <module name="org.jboss.weld.api"/>
        <module name="org.jboss.weld.core"/>
        <module name="org.jboss.weld.spi"/>
        <module name="org.jboss.staxmapper"/>
        <module name="org.jboss.vfs"/>
        <module name="org.wildfly.common"/>
        <module name="org.wildfly.extension.request-controller"/>
        <module name="org.wildfly.security.elytron-private"/>
        <module name="org.wildfly.transaction.client"/>
        <module name="java.xml"/>
        <module name="java.desktop"/>
    </dependencies>
</module>
```

Which depends on the `jberet` dep like this:

```xml
        <module name="org.jberet.jberet-core" services="import"/>
```

I can't see anything that has error in configuration, and have no idea on how to troubleshooting and fix the problem.

Could you please give some suggestions how to fix this? Thanks!

Reply:
I'm not too sure, but WildFly has a `batch-jberet` layer. We likely need to mimic that more than what we do for `resteasy` and `resteasy-spring` because those do not have layers. However, part of the layer is likely the subsystem so there might be more to it than that.

I'm not even sure a feature pack is needed here. All we really want to do is override the versions.